graphics/cursorman.cpp

```diff
@@ -106,57 +123,62 @@ void CursorManager::popAllCursors() {
 }
 
 void CursorManager::replaceCursor(const void *buf, uint w, uint h, int hotspotX, int hotspotY, uint32 keycolor, bool dontScale, const Graphics::PixelFormat *format, const byte *mask) {
+	PixelFormat pixelFormat;
+#ifdef USE_RGB_COLOR
+	if (format)
+		pixelFormat = *format;
+	else
+#endif
+		pixelFormat = PixelFormat::createFormatCLUT8();
+
+	Surface surf;
+	// we wont touch 'buf' ...
+	surf.init(w, h, w, const_cast<void *>(buf), pixelFormat);
+
+	replaceCursor(surf, hotspotX, hotspotY, keycolor, dontScale, mask);
+}
+
+void CursorManager::replaceCursor(const Surface &surf, int hotspotX, int hotspotY, uint32 keycolor, bool dontScale, const byte *mask) {
 	if (!g_system->hasFeature(OSystem::kFeatureCursorMask))
 		mask = nullptr;
 
 	if (_cursorStack.empty()) {
-		pushCursor(buf, w, h, hotspotX, hotspotY, keycolor, dontScale, format, mask);
+		pushCursor(surf, hotspotX, hotspotY, keycolor, dontScale, mask);
 		return;
 	}
 
 	Cursor *cur = _cursorStack.top();
 
-#ifdef USE_RGB_COLOR
-	uint size;
-	if (!format)
-		size = w * h;
-	else
-		size = w * h * format->bytesPerPixel;
-#else
-	uint size = w * h;
-#endif
+	const uint curSize = cur->_surf.w * cur->_surf.h * cur->_surf.format.bytesPerPixel;
+	const uint size = surf.w * surf.h * surf.format.bytesPerPixel;
 
-	if (cur->_size < size) {
-		delete[] cur->_data;
-		cur->_data = new byte[size];
-		cur->_size = size;
+	if (curSize < size) {
+		delete[] (byte *)cur->_surf.getPixels();
+		cur->_surf.setPixels(new byte[size]);
```

Comment:
You might want to keep the cur->_size variable rather than calculating curSize.
If the current cursor is replaced with a smaller size and then restored again with the previous, the surface with be allocated again unnecessarily.

A comment as to why surface create/free is not used here may also be useful.

Reply:
Good catch about `_size`, that can come in handy. Also added an explanatory comment about `Surface::create`.